common/src/Renderer/GLVertexAttributeType.h

```diff
@@ -20,296 +20,185 @@
 #ifndef TrenchBroom_AttributeSpec_h
 #define TrenchBroom_AttributeSpec_h
 
+#include "Ensure.h"
 #include "Macros.h"
 #include "Renderer/GL.h"
+#include "Renderer/ShaderProgram.h"
 
 #include <vecmath/vec.h>
 
 namespace TrenchBroom {
     namespace Renderer {
-        /**
-         * The values of this enum define the possible types of OpenGL vertex attributes.
-         */
-        enum class GLVertexAttributeTypeTag {
-            User,
-            Position,
-            Normal,
-            Color,
-            TexCoord0,
-            TexCoord1,
-            TexCoord2,
-            TexCoord3
+        enum class Normalize {
+            True, False
         };
 
-        /**
-         * Base template to define a vertex attribute type. This should be unused; use to the partial specializations
-         * to define actual vertex attribute types.
-         *
-         * @tparam T the type of the vertex attribute
-         * @tparam D the vertex component type
-         * @tparam S the number of components
-         */
-        template <GLVertexAttributeTypeTag T, GLenum D, size_t S>
-        class GLVertexAttributeType {
-        public:
-            using ComponentType = typename GLType<D>::Type;
-            using ElementType = vm::vec<ComponentType,S>;
-            static const size_t Size = sizeof(ElementType);
+        enum class AttributeNames {
+            // EntityLinkArrow
+            arrowPosition,
+            lineDir
+        };
 
-            /**
-             * Sets up a vertex buffer pointer for this attribute with the given index, stride, and offset.
-             *
-             * @param index the attribute's index (position in the vertices)
-             * @param stride the stride for the vertex buffer pointer
-             * @param offset the offset for the vertex buffer pointer
-             */
-            static void setup(const size_t index, const size_t stride, const size_t offset) {
-                unused(index);
-                unused(stride);
-                unused(offset);
+        inline const char* toString(const AttributeNames name) {
+            switch (name) {
+                case AttributeNames::arrowPosition: return "arrowPosition";
```

Comment:
I don't think these should be here. These names belong in the code that uses these classes. They should not know anything about them.

Reply:
What else could we do? The string has to be encoded in the type, and C++ doesn't allow strings as template params.

Reply:
I.e. what I mean by string encoded in the type is, note where I use this in EntityLinkRenderer.h:

`GLVertexAttributeUser<AttributeNames::arrowPosition, GL_FLOAT, 3, Normalize::False>`

Ideally this type would be:

`GLVertexAttributeUser<"arrowPosition", GL_FLOAT, 3, Normalize::False>`

but this isn't possible in C++ - so the AttributeNames enum class is just a workaround for this.

Reply:
You can pass a type that has a static member that holds the string, e.g.

```
struct Whatever {
    static const std::string name = "arrowPosition";
};
```

In that case, we do need separate types and not partial template specialization tho. But that's fine.

Reply:
Nice! I'll give that a try.